src/charm.py

```diff
@@ -61,366 +34,195 @@
+    def on_pebble_ready(self, event: EventBase) -> None:
+        """Run the pebble ready hook.
 
-    def on_database_joined(self, event):
-        self.provider.update_port("database", self.model.config["port"])
-        self.provider.update_address("database", self._bind_address())
+        Args:
+            event: the event object
+        """
+        if self._num_units() != self._goal_units():
+            self.unit.status = MaintenanceStatus("Waiting for units")
+            event.defer()
+            return
 
-    @status_catcher
-    def on_dashboard_joined(self, event):
-        if not self.unit.is_leader():
+        if self._set_config_file(event) is False:
             return
 
-        if isinstance(self.unit.status, BlockedStatus) and "dashboard" in self.unit.status.message:
-            self.unit.status = ActiveStatus()
+        if self._set_layer():
+            # self._container.restart("cassandra")
+            restart(self._container)
 
-        dashboard_tmpl = open(os.path.join(sys.path[0], "dashboard.json.tmpl"), "r").read()
+        if self.unit.is_leader():
+            if not self.cassandra.root_password(event):
+                return
 
-        self.dashboard_consumer.add_dashboard(dashboard_tmpl)
+        if len(self.model.relations["monitoring"]) > 0:
+            self._setup_monitoring()
 
-    def on_dashboard_broken(self, event):
-        self.dashboard_consumer.remove_dashboard()
-        if isinstance(self.unit.status, BlockedStatus) and "dashboard" in self.unit.status.message:
-            self.unit.status = ActiveStatus()
+        self.provider.update_address("database", self._hostname())
 
-    def on_dashboard_status_changed(self, event):
-        if event.valid:
-            self._dashboard_valid = True
-            self.unit.status = ActiveStatus()
-        elif event.error_message:
-            self._dashboard_valid = False
-            self.unit.status = BlockedStatus(event.error_message)
+        self.unit.status = ActiveStatus()
 
-    @status_catcher
-    def on_monitoring_joined(self, event):
-        self._setup_monitoring()
+    def on_config_changed(self, event: EventBase) -> None:
+        """Run the config changed hook.
 
-    def _reset_monitoring(self):
-        if not self.unit.is_leader():
+        Args:
+            event: the event object
+        """
+        heap_size = self.config["heap_size"]
+        if match(r"^\d+[KMG]$", heap_size, IGNORECASE) is None:
+            self.unit.status = BlockedStatus(f"Invalid Cassandra heap size setting: '{heap_size}'")
             return
-        if len(self.model.relations["monitoring"]) > 0:
-            for endpoint in self.prometheus_consumer.endpoints:
-                address, port = endpoint.split(":")
-                self.prometheus_consumer.remove_endpoint(address=address, port=int(port))
-            self._setup_monitoring()
+        if (
+            self._container.get_plan().to_dict().get("services", {}).get("cassandra")
```

Comment:
```suggestion
            "cassandra" in self._container.get_plan().services
```

I think the suggested would work (and be simpler), but needs confirming :)

Reply:
Yep looks right. I'm using the somewhat equivalent `if not self.container.get_plan().services.get(self._service_name):`

Reply:
Yup done.